src/tasks/tasks.hpp

```diff
@@ -0,0 +1,493 @@
+
+  TaskID operator|(const TaskID &other) const {
+    // calling this operator means you're building a TaskID to hold a dependency
+    TaskID result;
+    if (task != nullptr) result.dep.push_back(task);
+    result.dep.insert(result.dep.end(), dep.begin(), dep.end());
+    if (other.task != nullptr) result.dep.push_back(other.task);
+    result.dep.insert(result.dep.end(), other.dep.begin(), other.dep.end());
+    return result;
+  }
+
+  const std::vector<Task *> &GetIDs() const { return std::cref(dep); }
+
+  bool empty() const { return (!task && dep.size() == 0); }
+  Task *GetTask() { return task; }
+
+ private:
+  Task *task = nullptr;
+  std::vector<Task *> dep;
+};
+
+class Task {
+ public:
+  Task() = default;
+  template <typename TID>
+  Task(TID &&dep, const std::function<TaskStatus()> &func,
+       std::pair<int, int> limits = {1, 1})
+      : f(func), exec_limits(limits) {
+    if (dep.GetIDs().size() == 0 && dep.GetTask()) {
+      dependencies.insert(dep.GetTask());
+    } else {
+      for (auto &d : dep.GetIDs()) {
+        dependencies.insert(d);
+      }
+    }
+    // always add "this" to repeat task if it's incomplete
+    dependent[static_cast<int>(TaskStatus::incomplete)].push_back(this);
+  }
+
+  TaskStatus operator()() {
+    auto status = f();
+    if (task_type == TaskType::completion) {
+      // keep track of how many times it's been called
+      num_calls += (status == TaskStatus::iterate || status == TaskStatus::complete);
+      // enforce minimum number of iterations
+      if (num_calls < exec_limits.first && status == TaskStatus::complete)
+        status = TaskStatus::iterate;
+      // enforce maximum number of iterations
+      if (num_calls == exec_limits.second) status = TaskStatus::complete;
+    }
+    // save the status in the Task object
+    SetStatus(status);
+    return status;
+  }
+  TaskID GetID() { return this; }
+  bool ready() {
+    // check that no dependency is incomplete
+    bool go = true;
+    for (auto &dep : dependencies) {
+      go = go && (dep->GetStatus() != TaskStatus::incomplete);
+    }
+    return go;
+  }
+  void AddDependency(Task *t) { dependencies.insert(t); }
+  std::unordered_set<Task *> &GetDependencies() { return dependencies; }
+  void AddDependent(Task *t, TaskStatus status) {
+    dependent[static_cast<int>(status)].push_back(t);
+  }
+  std::vector<Task *> &GetDependent(TaskStatus status = TaskStatus::complete) {
+    return dependent[static_cast<int>(status)];
+  }
+  void SetType(TaskType type) { task_type = type; }
+  TaskType GetType() { return task_type; }
+  void SetStatus(TaskStatus status) {
+    std::lock_guard<std::mutex> lock(mutex);
+    task_status = status;
+  }
+  TaskStatus GetStatus() {
+    std::lock_guard<std::mutex> lock(mutex);
+    return task_status;
+  }
+  void reset_iteration() { num_calls = 0; }
+
+ private:
+  std::function<TaskStatus()> f;
+  std::array<std::vector<Task *>, 3> dependent;
```

Comment:
What is `dependent` supposed to store? Would having `auto *dep_task = dependent[TaskStatus::complete][3]` imply that `dep_task` requires that the `task_status` of `this` is `TaskStatus::complete`?

Reply:
It stores a vector of tasks that might possibly be ready to run for a given task status returned by the owning Task.  So, for example, every Task points at itself for `TaskStatus::incomplete` so that if a task returns that status, it gets added to the queue again.  For `TaskStatus::complete`, the vector of tasks is just all the tasks that have this Task as a dependency.  Does that make sense?

Reply:
yep, that makes sense, thanks. A brief comment in the code might be helpful in the future.

Reply:
done